ethcore/src/account.rs

```diff
@@ -364,6 +384,42 @@ impl Account {
 		stream.append(self.code_hash.as_ref().unwrap_or(&SHA3_EMPTY));
 		stream.out()
 	}
+
+	/// Clone basic account data
+	pub fn clone_basic(&self) -> Account {
+		Account {
+			balance: self.balance.clone(),
+			nonce: self.nonce.clone(),
+			storage_root: self.storage_root.clone(),
+			storage_cache: Self::empty_storage_cache(),
+			storage_changes: HashMap::new(),
+			code_hash: self.code_hash.clone(),
+			code_size: self.code_size.clone(),
+			code_cache: Bytes::new(),
+			filth: self.filth,
+			address_hash: self.address_hash.clone(),
+		}
+	}
+
+	/// Clone account data and dirty storage keys
+	pub fn clone_dirty(&self) -> Account {
+		let mut account = self.clone_basic();
+		account.storage_changes = self.storage_changes.clone();
+		account
+	}
+
+	/// Replace self with the data from other account merging storage cache
+	pub fn merge_with(&mut self, mut other: Account) {
+		assert!(self.storage_is_clean());
+		assert!(other.storage_is_clean());
+		let mut our_cache = replace(&mut self.storage_cache, Self::empty_storage_cache()).into_inner();
+		let updated_cache = replace(&mut other.storage_cache, Self::empty_storage_cache()).into_inner();
+		for (k, v) in updated_cache.into_iter() {
+			our_cache.insert(k.clone() , v.clone()); //TODO: cloning should not be required here
+		}
+		*self = other;
+		self.storage_cache = RefCell::new(our_cache);
```

Comment:
seems a little over-complex in terms of assignments; why do `our_cache` and `updated_cache` need pulling out only to be replaced here? why reassign self?